File: hnswlib-spark/src/main/java/org/github/jelmer/knn/spark/DistanceFunctions.java
package org.github.jelmer.knn.spark;

import org.apache.spark.ml.linalg.Vector;

public class DistanceFunctions {

    public static double cosineDistance(Vector u, Vector v) {
        if (u.size() != v.size()) {
            throw new IllegalArgumentException("Vectors have non-matching dimensions");
        }

        double dot = 0.0f;
        double nru = 0.0f;
        double nrv = 0.0f;
        for (int i = 0; i < u.size(); i++) {
            dot += u.apply(i) * v.apply(i);
            nru += u.apply(i) * u.apply(i);
            nrv += v.apply(i) * v.apply(i);
        }

        double similarity = dot / (Math.sqrt(nru) * Math.sqrt(nrv));
        return 1 - similarity;
    }


    public static double innerProduct(Vector u, Vector v) {
        if (u.size() != v.size()) {
            throw new IllegalArgumentException("Vectors have non-matching dimensions");
        }

        double dot = 0;
        for (int i = 0; i < u.size(); i++) {
            dot += u.apply(i) * v.apply(i);
        }

        return 1 - dot;
    }
}

File: hnswlib-spark/src/main/scala/com/github/jelmerk/spark/knn/DistanceFunctions.scala
package com.github.jelmerk.spark.knn

import org.apache.spark.ml.linalg.SparseVector

object DistanceFunctions {

  def cosineDistance(u: SparseVector, v: SparseVector): Double = {
    val denom = norm(u) * norm(v)
    val dot = innerProduct(u, v)

    if (denom == 0d) 1d
    else 1 - dot / denom
  }

  def innerProductDistance(u: SparseVector, v: SparseVector): Double = 1 - innerProduct(u, v)

  private def norm(u: SparseVector): Double = math.sqrt(u.values.map(v => v * v).sum)

  private def innerProduct(u: SparseVector, v: SparseVector): Double = {
    val uIndices = u.indices
    val vIndices = v.indices

    val uValues = u.values
    val vValues = v.values

    var dot = 0d

    var i = 0
    var j = 0

    while(i < uIndices.length && j < vIndices.length) {
      if (uIndices(i) < vIndices(j)) {
        i += 1
      } else if (uIndices(i) > vIndices(j)) {
        j += 1
      } else {
        dot += uValues(i) * vValues(j)
        i += 1
        j += 1
      }
    }
    dot
  }

}
